docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
+  pods to query for metrics within their own namespace.
+
+- separating querying from pushing: we should allow certain users to push, and
+  certain users to query, but not necessarily tie the two together
+
+- allowing different push scope permissions: we should be able to specify that
+  some users are only allowed to push metrics for the pods within their own
+  RC or namespace, while others should be allowed to push for any object in
+  the cluster.
+
+Proposed Solution
+-----------------
+
+Like in Kubernetes, authentication and authorization should be split into
+separate interfaces.  Each would be configured using a URL flag at the command
+line, similarly to how sinks and sources are currently configured.  Adding
+multiple authentication drivers would be supported (for example, to support
+authenticating from both client certificates and bearer tokens).
+
+### Authentication ###
+
+For authentication, the Kubernetes interface
+(`"k8s.io/kubernetes/pkg/auth/authenticator".Request`) can be reused so that
+any authentication plugins for Kubernetes can be easily used with Heapster.
+The interface is generic enough that it could also be used for specialized
+Heapster-specific interfaces.
+
+For example, if it was desired to use client certificate authentication,
+the command line flag might look like
+`--authn=clientca:/etc/heapster/ca.crt`.
+
+### Authorization ###
+
+For authorization, a system tailored to Heapster directly would be more useful:
+
+```go
+// This is basically just core.HistoricalKey from the Oldtimer PR,
+// renamed to be more generic
+type MetricKey struct {
```

Comment:
Do you really need all this information?  Excessive information makes it hard to cache authorization decisions, and makes it harder to integrate with a general authorization system like RBAC.